core/server/worker/src/main/java/alluxio/worker/block/allocator/Allocator.java

```diff
@@ -68,10 +69,10 @@ public static Allocator create(BlockMetadataView view) {
    * @param blockSize the size of block in bytes
    * @param location the location in block store
    * @param view of the block metadata
-   * @param skipReview whether the review should be skipped
+   * @param reviewFunc the {@link Reviewer} logic that rejects bad block allocations
```

Comment:
By using a `reviewFunc` lambda, `skipReview` can be replaced. Whether to skip can be determined by `AllocationCoordinator` on one level above.

Reply:
Although the Allocator is decoupled from specific Reviewer implementation (not holding the instance anymore), admittedly this is not 100% decoupled. This is due to 2 reasons:

1. Java DOES NOT have "generator" like python `yield` grammar. I have to generate the full list of possible allocation proposals and pass it back to the `AllocationCoordinator` to review. There's no "Allocator saying I have this proposal, Reviewer rejects so Allocator you think of one more" interactive allocation decision.

1. Even if we implement passing the list of proposals to the `AllocationCoordinator` and select one, it's hard to update the stateful `Allocator` after `AllocationCoordinator` makes a decision. `RoundRobinAllocator` is stateful and needs to know the decision in order to make the next one. `RoundRobinAllocator` has been the source of evil in a few different locations :(

Reply:
@gpang With above being said, I currently don't have a better idea than the implementation im putting here. See if this is acceptable to you.

Reply:
Well, if we organize the StorageDir in a LinkedList we can probably have a "reject and gimme next proposal" implementation in the `AllocationCoordinator`.

Reply:
If this implementation is not clean (not decoupling enough) for you, there are other options:

1. There are not many StorageDir. It hurts very little to just order them and pass the list back from the `Allocator`. It's just a bit hard to update the Allocator with the allocation decision, but that is also feasible.
1. If we order `StorageDir` in the Allocator as a LinkedList, we can do a "reject and Allocator gimme your next proposal" within the `AllocationCoordinator`.

Reply:
I see, so the main issue is that the allocator needs to update internal state, depending on the decision by the reviewer.

However, would changing the api to basically return an iterator be too big of a change? basically it could look something like:

```
it = allocator.getNewIterator();

while (it.hasNext()) {
  if (reviewer.accept(it.next()) {
    // This storage dir is accepted!
  }
}
// NO storage dir possible
```

In this model, all the internal state is managed within the iterator implementation, for each allocator.

Could something like this work?